testing/upgradetest/upgrader.go

```diff
@@ -0,0 +1,358 @@
+	var upgradeOpts upgradeOpts
+	upgradeOpts.skipVerify = true
+	for _, o := range opts {
+		o(&upgradeOpts)
+	}
+
+	// ensure that both the starting and ending fixtures are prepared
+	err := startFixture.EnsurePrepared(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to prepare the startFixture: %w", err)
+	}
+	err = endFixture.EnsurePrepared(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to prepare the endFixture: %w", err)
+	}
+
+	// start fixture gets the agent configured to use a faster watcher
+	err = ConfigureFastWatcher(ctx, startFixture)
+	if err != nil {
+		return fmt.Errorf("failed configuring the start agent with faster watcher configuration: %w", err)
+	}
+
+	// get the versions from each fixture (that ensures that it's always the
+	// same version that the fixture is working with)
+	startVersionInfo, err := startFixture.ExecVersion(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get start agent build version info: %w", err)
+	}
+	startParsedVersion, err := version.ParseVersion(startVersionInfo.Binary.String())
+	if err != nil {
+		return fmt.Errorf("failed to get parsed start agent build version (%s): %w", startVersionInfo.Binary.String(), err)
+	}
+	endVersionInfo, err := endFixture.ExecVersion(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get end agent build version info: %w", err)
+	}
+
+	if upgradeOpts.preInstallHook != nil {
+		if err := upgradeOpts.preInstallHook(); err != nil {
+			return fmt.Errorf("pre install hook failed: %w", err)
+		}
+	}
+
+	logger.Logf("Installing version %q", startParsedVersion.VersionWithPrerelease())
+
+	// install the start agent
+	var nonInteractiveFlag bool
+	if Version_8_2_0.Less(*startParsedVersion) {
+		nonInteractiveFlag = true
+	}
+	installOpts := atesting.InstallOpts{
+		NonInteractive: nonInteractiveFlag,
+		Force:          true,
+	}
+	output, err := startFixture.Install(ctx, &installOpts)
+	if err != nil {
+		return fmt.Errorf("failed to install start agent (err: %w) [output: %s]", err, string(output))
+	}
+
+	if upgradeOpts.postInstallHook != nil {
+		if err := upgradeOpts.postInstallHook(); err != nil {
+			return fmt.Errorf("post install hook failed: %w", err)
+		}
+	}
+
+	// wait for the agent to be healthy and correct version
+	err = WaitHealthyAndVersion(ctx, startFixture, startVersionInfo.Binary, 2*time.Minute, 10*time.Second, logger)
+	if err != nil {
+		// context added by WaitHealthyAndVersion
+		return err
+	}
+
+	if upgradeOpts.preUpgradeHook != nil {
+		if err := upgradeOpts.preUpgradeHook(); err != nil {
+			return fmt.Errorf("pre upgrade hook failed: %w", err)
+		}
+	}
+
+	logger.Logf("Upgrading from version %q to version %q", startParsedVersion, endVersionInfo.Binary.String())
+
+	upgradeCmdArgs := []string{"upgrade", endVersionInfo.Binary.String()}
+	if upgradeOpts.sourceURI != "" {
+		// specific ---source-uri
+		upgradeCmdArgs = append(upgradeCmdArgs, "--source-uri", upgradeOpts.sourceURI)
+	} else if upgradeOpts.customPgp == nil {
```

Comment:
Not sure why this is an `else if` clause instead of being a simple `if`... 
Are we saying that this function will ignore PGP options if a custom `source-uri` is provided ? 
Shouldn't that decision be left to the test calling this function with the appropriate options?

Reply:
True. I will update.

Reply:
Updated.